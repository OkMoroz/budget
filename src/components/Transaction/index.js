import PropTypes from "prop-types";
import { useContext, useCallback, memo } from "react";

import { AppContext } from "../../providers/context";
import {
  Wrapper,
  TransactionDate,
  Value,
  Comment,
  Icon,
  Category,
  DeleteButton,
} from "./styles";
import Star from "../../assets/img/star.svg";
import StarFilled from "../../assets/img/star_filed.svg";

const Transaction = memo(
  ({
    transaction: { id, value, date, category, comment, isStarred },
    onDelete, 
    onStarClick,
  }) => {
    const { state } = useContext(AppContext);

    const deleteItem = useCallback(() => onDelete(id), [id]);

    return (
      <Wrapper value={value}>
        <Icon onClick={() => onStarClick(id)}>
          <img src={isStarred ? StarFilled : Star} />
        </Icon>

        <TransactionDate>{date}</TransactionDate>
        <Value>
          {value.toFixed(2)}, {state.currency}
        </Value>
        <Category>{category}</Category>
        <Comment>{comment}</Comment>
        <DeleteButton onClick={deleteItem}>DELETE</DeleteButton>
      </Wrapper>
    );
  }
);

Transaction.propTypes = {
  transaction: PropTypes.shape({
    id: PropTypes.number, 
    value: PropTypes.number,
    date: PropTypes.string,
    category: PropTypes.string,
    comment: PropTypes.string,
    isStarred: PropTypes.bool,
  }),
  onDelete: PropTypes.func,
  onStarClick: PropTypes.func,
};

Transaction.defaultProps = {
  transaction: {
    label: "",
    value: 0,
  },
};

export default Transaction;
