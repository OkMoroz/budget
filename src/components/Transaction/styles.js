import styled from "styled-components";

export const Wrapper = styled.div`
  background: ${({ value }) => (value < 0 ? "#e2e1e1" : "#fff")};
  border: 1px solid #d9b3ff;
  border-radius: 3px;
  padding: 5px;
  margin-bottom: 20px;
  display: flex;
  align-items: center;
  justify-content: space-between;
`;
Wrapper.displayName = "TransactionWrapper";

export const TransactionDate = styled.div`
  flex-grow: 1;
`;
TransactionDate.displayName = "TransactionDate";

export const Value = styled.div`
  flex-grow: 1;
`;
Value.displayName = "Value";

export const Category = styled.div`
  flex-grow: 2;
`;
Category.displayName = "Comment";

export const Comment = styled.div`
  flex-grow: 3;
`;
Comment.displayName = "Comment";

export const Icon = styled.span`
  img {
    max-width: 20px;
    margin-right: 10px;
  }
`;
Icon.displayName = "Icon";

export const DeleteButton = styled.button`
  background-color: #f4f4f4;
  color: #333;
  border: none;
  border-radius: 2px;
  padding: 5px 10px;
  cursor: pointer;

  &:hover {
    background-color: #d9b3ff;
    color: #ff868e;
    font-weight: 700;
  }

  &:active {
    background-color: #ddd;
  }
`;
DeleteButton.displayName = "DeleteButton";
