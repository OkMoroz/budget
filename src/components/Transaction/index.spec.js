import React from "react";
import { render } from "@testing-library/react";
import Transaction from ".";

describe("Transaction component", () => {
  let props;

  beforeEach(() => {
    props = {
      transaction: {
        value: 23,
        comment: "test",
        date: "01.01.2020",
      },
    };
  });

  it("should show transaction", () => {
    const { container } = render(<Transaction {...props} />);
    expect(container).toMatchSnapshot();
  });

  it("should show 2 zeros after amount", () => {
    const { getByTestId } = render(<Transaction {...props} />);
    const valueElement = getByTestId("transaction-value");
    expect(valueElement.textContent).toBe("23.00");
  });
});
