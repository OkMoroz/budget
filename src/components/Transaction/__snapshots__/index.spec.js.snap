// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transaction component should show transaction 1`] = `
<div>
  <div
    class="sc-bcPKhP ivTuzb"
    value="23"
  >
    <div
      class="sc-grXZZQ kxISmc"
    >
      01.01.2020
    </div>
    <div
      class="sc-dkjKgF gZdpGr"
      data-testid="transaction-value"
    >
      23.00
    </div>
    <div
      class="sc-hLclGa jRXliq"
    >
      test
    </div>
  </div>
</div>
`;
